src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIteratorExtensions.cs

```diff
@@ -232,6 +232,32 @@ public static ArraySegment<byte> GetArraySegment(this MemoryPoolIterator start,
             return new ArraySegment<byte>(array, 0, length);
         }
 
+        public static ArraySegment<byte> PeekArraySegment(this MemoryPoolIterator iter)
+        {
+            if (iter.IsDefault || iter.IsEnd)
+            {
+                return default(ArraySegment<byte>);
+            }
+
+            if (iter.Index < iter.Block.End)
+            {
+                return new ArraySegment<byte>(iter.Block.Array, iter.Index, iter.Block.End - iter.Index);
+            }
+
+            var block = iter.Block.Next;
+            while (block != null)
+            {
+                if (block.Start < block.End)
+                {
+                    return new ArraySegment<byte>(block.Array, block.Start, block.End - block.Start);
+                }
+                block = block.Next;
+            }
+
+            // The following should be unreachable due to the IsEnd check above
+            throw new InvalidOperationException("This should be unreachable!");
```

Comment:
No, I was just remarking that this is the classic "can't happen" scenario (http://www.catb.org/jargon/html/C/can-t-happen.html) 😄